### PR TITLE
runtime: blacklist /dev, /proc and /sys from watch dirs.

### PR DESCRIPTION
**Pull request by htuch**
server_test proto fuzzing resulted in OOM because of /dev traversal. It
seems bad to allow runtime paths to point in these places.

Risk Level: Low
Testing: New unit tests.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/runtime/runtime_impl.cc [hunk 1]
```diff
@@ -231,6 +231,10 @@ DiskLayer::DiskLayer(const std::string& name, const std::string& path,
 
 void DiskLayer::walkDirectory(const std::string& path, const std::string& prefix) {
   ENVOY_LOG(debug, "walking directory: {}", path);
+  // Check if this is an obviously bad path.
```
Comment (mattklein123) [inline source/common/runtime/runtime_impl.cc]: Per offline convo, I still think the correct solution here is a max recursion check. Can we TODO that? Additionally, It seems like we should potentially do this check in any place we attempt to load a file (TLS certs, etc.)? Should we issue/TODO that also?

Comment (htuch) [reply source/common/runtime/runtime_impl.cc]: Agree. Added TODO for recursion depth and filed https://github.com/envoyproxy/envoy/issues/3170.

File: source/common/filesystem/filesystem_impl.cc [hunk 2]
```diff
@@ -61,6 +63,36 @@ std::string fileReadToEnd(const std::string& path) {
   return file_string.str();
 }
 
+std::string canonicalPath(const std::string& path) {
+  // TODO(htuch): When we are using C++17, switch to std::filesystem::canonical.
+  char* resolved_path = ::realpath(path.c_str(), nullptr);
+  if (resolved_path == nullptr) {
+    throw EnvoyException(fmt::format("Unable to determine canonical path for {}", path));
```
Comment (mattklein123) [inline source/common/filesystem/filesystem_impl.cc]: Can you cover this case? Perhaps if you just pass realpath() a bogus path? (I'm not sure what the OSX issue was).

File: source/common/runtime/runtime_impl.h [hunk 3]
```diff
@@ -143,10 +143,12 @@ class DiskLayer : public OverrideLayerImpl, Logger::Loggable<Logger::Id::runtime
     DIR* dir_;
   };
 
-  void walkDirectory(const std::string& path, const std::string& prefix);
+  void walkDirectory(const std::string& path, const std::string& prefix, uint32_t depth);
 
   const std::string path_;
   Api::OsSysCalls& os_sys_calls_;
+  // Maximum recursion depth for walkDirectory().
+  const uint32_t MaxWalkDepth = 4;
```
Comment (mattklein123) [inline source/common/runtime/runtime_impl.h]: I would go higher here. This might even be about where we are with our normal topology. Maybe 20?

**Pull request merged**